diff --git a/Global/Include.h b/Global/Include.h
index c77105c..5cd471a 100644
--- a/Global/Include.h
+++ b/Global/Include.h
@@ -2,11 +2,13 @@
 #define __STANDARD_DEF__
 #include <stdint.h>
 
-#ifndef DEBUG
-#define DEBUG 1
-void __error__(char *pcFilename, uint32_t ui32Line);
+#ifdef __cplusplus
+extern "C"
+{
 #endif
 
+void __error__(char *pcFilename, uint32_t ui32Line);
+
 extern void
 UARTprintf(const char *pcString, ...);
 #define printf UARTprintf
@@ -86,4 +88,9 @@ static inline unsigned int GET_MASK_POS(unsigned int MASK)
 #define BIT29   (1<<29)
 #define BIT30   (1<<30)
 #define BIT31   (1<<31)
+
+#ifdef __cplusplus
+}
+#endif
+
 #endif
diff --git a/INC/device/eeprom.h b/INC/device/eeprom.h
index 5525b9c..fb9b96f 100644
--- a/INC/device/eeprom.h
+++ b/INC/device/eeprom.h
@@ -2,6 +2,11 @@
 #define _EEPROM_H_
 #include <stdint.h>
 
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
 enum eepromTransferMode {
   I2CMode,
   SPIMode,
@@ -10,4 +15,9 @@ enum eepromTransferMode {
 
 void eepromWrite(uint32_t writeAddr, uint32_t size, uint8_t *wData);
 void eepromRead(uint32_t readAddr, uint32_t size, uint8_t *rData);
+
+#ifdef __cplusplus
+}
+#endif
+
 #endif
\ No newline at end of file
diff --git a/INC/gpio.h b/INC/gpio.h
index 5914267..4c968bc 100644
--- a/INC/gpio.h
+++ b/INC/gpio.h
@@ -1,6 +1,11 @@
 #ifndef __GPIO_H__
 #define __GPIO_H__
 
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
 #define GPIOPA_APB_BASE  0x40004000
 #define GPIOPB_APB_BASE  0x40005000
 #define GPIOPC_APB_BASE  0x40006000
@@ -25,4 +30,9 @@
 
 // APIs
 void PAD_GPIOI2CPinConfig(void);
+
+#ifdef __cplusplus
+}
+#endif
+
 #endif
diff --git a/INC/gtimer_driver.h b/INC/gtimer_driver.h
index 7f1b78d..49ac7fd 100644
--- a/INC/gtimer_driver.h
+++ b/INC/gtimer_driver.h
@@ -1,6 +1,11 @@
 #ifndef _TIMER_DRIVER_
 #define _TIMER_DRIVER_
 
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
 enum {
     PIOSC_CLOCK,
     SYS_CLOCK,
@@ -15,4 +20,9 @@ int SysTickIsCounting(void);
 unsigned int SysTickCurrentGet(void);
 void SysTickPeriodSet(unsigned int period);
 unsigned int SysTickPeriodGet(void);
+
+#ifdef __cplusplus
+}
+#endif
+
 #endif
diff --git a/INC/i2c.h b/INC/i2c.h
index 9990a7e..fb9ca86 100644
--- a/INC/i2c.h
+++ b/INC/i2c.h
@@ -2,6 +2,12 @@
 #define __I2C_HW__
 #include <stdint.h>
 #include <stdbool.h>
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
 // Master Registers
 #define I2C_BASE 0x40020000
 #define I2CMSA_REG      *(volatile unsigned int*)(I2C_BASE    + 0x0)   // M/Slave Address                  (RW)
@@ -39,4 +45,9 @@ void I2C_Init(void);
 bool I2C_WriteBytes(uint8_t slaveAddress, uint32_t count, uint8_t *data);
 bool I2C_ReadBytes(uint32_t slaveAddress, uint32_t count, uint8_t *data);
 // Slave Control APIs
+
+#ifdef __cplusplus
+}
+#endif
+
 #endif
diff --git a/INC/spi.h b/INC/spi.h
index 87bf161..2982e7d 100644
--- a/INC/spi.h
+++ b/INC/spi.h
@@ -3,6 +3,11 @@
 #include <stdint.h>
 #include <stdbool.h>
 
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
 #define SPI_BASE 0x40008000
 
 #define SPI_SSICR0         *(volatile unsigned int*) (SPI_BASE + 0x000)        // RW    0x0000.0000 SSI Control 0
@@ -32,4 +37,9 @@
 //APIs
 bool SPI_WriteBytes(uint32_t count, uint8_t *data);
 bool SPI_ReadBytes(uint32_t count, uint8_t *data);
+
+#ifdef __cplusplus
+}
+#endif
+
 #endif
\ No newline at end of file
diff --git a/INC/sys.h b/INC/sys.h
index 8cde87f..20854a2 100644
--- a/INC/sys.h
+++ b/INC/sys.h
@@ -2,6 +2,12 @@
 #define __SYS_H__
 #include <stdint.h>
 #include "Global/Include.h"
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
 // Register Map//
 #define SYSCTL_BASE 0x400FE000
 
@@ -133,4 +139,9 @@
 // APIs
 void PAD_SysPeripheralClockEnable(uint32_t regBase);
 void PAD_SysClockSet(void);
+
+#ifdef __cplusplus
+}
+#endif
+
 #endif
diff --git a/projects/i2cprj/main.c b/projects/i2cprj/main.cpp
similarity index 87%
rename from projects/i2cprj/main.c
rename to projects/i2cprj/main.cpp
index 171bc51..3fde973 100644
--- a/projects/i2cprj/main.c
+++ b/projects/i2cprj/main.cpp
@@ -35,15 +35,38 @@
 
 #define AT24C64_ADDR 0x50
 uint8_t cData[8192] = {0};
-#if 1
+
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
 void
 __error__(char *pcFilename, uint32_t ui32Line)
 {
     UARTprintf("[DEBUG] %s:%d\r\n", pcFilename, ui32Line);
 //    while(1);
 }
+#ifdef __cplusplus
+}
 #endif
 
+class testOut {
+    public:
+        int a;
+        int b;
+    void test(void) {
+        printf("[]\r\n");
+        test(1,2);
+    }
+    private:
+        int c;
+        int d;
+    void test(int a, int b) {
+        printf("[call from]\r\n");
+    }
+};
+
 void InitConsole()
 {
     SysCtlPeripheralEnable(SYSCTL_PERIPH_GPIOA);
@@ -74,10 +97,11 @@ void eepromDump(uint32_t size, uint8_t * dData) {
 }
 
 
-void main(void){
+int main(void){
     uint8_t testbuff[8] = {0xaa,0xaa};
     uint8_t mulbytes[8] = {0x8,0x9,0x0a,0xb,0xc,0xd,0xe,0xf};
     uint32_t i = 256;
+    testOut tempObj;
     // while(--i) cData[i] = 0xAA;
 
 #if defined(gcc)
@@ -88,7 +112,7 @@ void main(void){
     InitConsole();
     I2C_Init();
     memset(cData, 0xAA, sizeof(uint8_t)*8192);
-
+    tempObj.test();
     // I2C Eeprom
     // eepromDump(8192, cData);
     // eepromRead(0x20, 3, testbuff);
@@ -104,4 +128,5 @@ void main(void){
 
     UARTprintf("Clock=%d\r\n", SysCtlClockGet());
     while(1);
+    return 0;
 }
diff --git a/projects/i2cprj/mkdefflags b/projects/i2cprj/mkdefflags
index b1f7ff8..63bfb24 100644
--- a/projects/i2cprj/mkdefflags
+++ b/projects/i2cprj/mkdefflags
@@ -4,6 +4,7 @@ endif
 
 #Toolchains
 CC = arm-none-eabi-gcc
+CXX = arm-none-eabi-gcc
 OBJCOPY = arm-none-eabi-objcopy
 LD = arm-none-eabi-ld
 AR = arm-none-eabi-ar
@@ -44,6 +45,19 @@ CFLAGS=-mthumb             \
 	-c                  
 	# --print-multi-lib
 
+CXXFLAGS=-mthumb             \
+	${CPU}              \
+	${FPU}              \
+	-ffunction-sections \
+	-fdata-sections     \
+	-fno-exceptions     \
+	-MD                 \
+	-std=c++98            \
+	-Wall               \
+	-pedantic           \
+	-DPART_${MCU}       \
+	-c                  
+	# --print-multi-lib
 
 #
 # The flags passed to the linker.
@@ -83,4 +97,5 @@ LIBM:=${shell ${CC} ${CFLAGS} -print-file-name=libm.a}
 # Add the include file paths to AFLAGS and CFLAGS.
 #
 AFLAGS+=${patsubst %,-I%,${subst :, ,${INCLUDE_PATH}}}
-CFLAGS+=${patsubst %,-I%,${subst :, ,${INCLUDE_PATH}}}
\ No newline at end of file
+CFLAGS+=${patsubst %,-I%,${subst :, ,${INCLUDE_PATH}}}
+CXXFLAGS+=${patsubst %,-I%,${subst :, ,${INCLUDE_PATH}}}
\ No newline at end of file
diff --git a/projects/i2cprj/mkdefs b/projects/i2cprj/mkdefs
index ed7200c..7305282 100644
--- a/projects/i2cprj/mkdefs
+++ b/projects/i2cprj/mkdefs
@@ -7,6 +7,15 @@ ${OUTDIR}/%.o: %.c
 	  ${CC} ${CFLAGS} -D${COMPILER} -o $@ $< ;          \
 	fi
 
+${OUTDIR}/%.o: %.cpp
+	if [ -d "${OUTDIR}" ];                         \
+	 then                                            \
+	  ${CXX} ${CXXFLAGS} -D${COMPILER} -o $@ $<;          \
+	 else                                            \
+	  ${MKDIR} ${OUTDIR}           ;                  \
+	  ${CXX} ${CXXFLAGS} -D${COMPILER} -o $@ $< ;          \
+	fi
+
 ${OUTDIR}/%.a:
 	$(AR) -cr $@ $^
 
